compiler/vm/vmchecks.nim

```diff
@@ -0,0 +1,130 @@
+## This module implements the handle validation logic.
+##
+## Handles are validate by first making sure that they point to VM owned
+## memory. After finding the allocation the handle points into, the
+## allocation's layout is walked in order to verify that there exists a valid
+## location of the handle's type at the offset relative to the start of
+## the allocation.
+##
+## Special handling is required for variant objects, since whether a
+## sub-location is valid depends on the run-time value of the respective
+## disciminator(s) there.
+
+import
+  compiler/vm/[
+    vmdef,
+    vmmemory
+  ]
+
+from compiler/vm/vmobjects import variantFieldIndices
+from compiler/vm/vmtypes import alignedSize
+
+func searchInObject(h: MemRegionPtr, otyp, t: PVmType, o: int): AccessViolationReason
+func testLocationType(locType: PVmType, t: PVmType): AccessViolationReason
+
+func searchInArray(h: MemRegionPtr, etyp: PVmType, t: PVmType, len, stride, o: int): AccessViolationReason =
+  let i = o /% stride
+  assert i < len
+
+  if o == i * stride:
+    testLocationType(etyp, t)
+  else:
+    case etyp.kind
+    of realAtomKinds:
+      avrNoLocation
+    of akObject:
+      searchInObject(h.getSubHandle(i*stride), etyp, t, o - (i*stride))
+    of akArray:
+      searchInArray(h.getSubHandle(i*stride), etyp.elementType, t, etyp.elementCount, etyp.elementStride, o - (i*stride))
+
+
+func searchInObject(h: MemRegionPtr, otyp, t: PVmType, o: int): AccessViolationReason =
+  assert uint(o) < otyp.sizeInBytes
+  assert otyp.kind == akObject
+  if otyp.branches.len == 0:
+    for (i, f) in otyp.objFields.pairs:
+      if f.offset == o:
+        return testLocationType(f.typ, t)
+      elif o < f.offset + int(f.typ.sizeInBytes):
+        case f.typ.kind
+        of akObject:
+          return searchInObject(h.getSubHandle(f.offset), f.typ, t, o - f.offset)
+        of akArray:
+          return searchInArray(h.getSubHandle(f.offset), f.typ.elementType, t, f.typ.elementCount, f.typ.elementStride, o - f.offset)
+        else:
+          return avrNoLocation
+  else:
+    for i in variantFieldIndices(h, otyp, 0):
```

Comment:
Create an iterator for this and then both branches unify under them, from what I can see.